File: node/mods/rustscript/lib/opcodes/checkhash.js
/**
 * Purpose: CHECKHASH opcode — verify witness preimage hashes to script hash.
 */

module.exports = {
  name: 'CHECKHASH',
  description: 'Verify that a preimage hashes to a given Blake3 hash.',
  exampleScript: {
    op: 'CHECKHASH',
    hash: '<hash>',
    witness: {
      input: '<secret>'
    }
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }
    if (!context.app || !context.app.crypto || typeof context.app.crypto.hash !== 'function') {
      return false;
    }

    const witness = node.witness;
    if (!witness || typeof witness !== 'object' || Array.isArray(witness)) {
      return false;
    }

    if (!Object.prototype.hasOwnProperty.call(witness, 'input')) {
      return false;
    }

    const input = witness.input;
    if (input === undefined || input === null) {
      return false;
    }
    if (typeof input === 'string' && input.length === 0) {
      return false;
    }

    const hash = node.hash;
    if (typeof hash !== 'string' || hash.length === 0) {
      return false;
    }

    const computed = context.app.crypto.hash(input);
    if (typeof computed !== 'string' || computed.length === 0) {
      return false;
    }

    return computed === hash;
  }
};

File: rust/saito-core/src/core/consensus/scripting/opcodes/checkhash.rs
<<<<<<< HEAD
use crate::core::consensus::block::Block;
use crate::core::consensus::transaction::Transaction;
use crate::core::defs::PrintForLog;
use crate::core::util::crypto;
use serde_json::Value;

pub struct CheckHash {
    pub name: String,
    pub description: String,
    pub script: String,
=======
//! CHECKHASH opcode behavior.
//!
//! Eventually: `script.validate()` → `CheckHash::execute(script, node)`.

use serde_json::Value;

use super::super::Script;

/// CHECKHASH opcode definition and execution entry point.
pub struct CheckHash {
    /// Opcode name.
    pub name: String,
    /// Human-readable description.
    pub description: String,
    /// Example JSON script.
    pub script: String,
    /// Schema definition.
>>>>>>> d0f828fa (fix: rustscript rust implementation)
    pub schema: Value,
}

impl CheckHash {
<<<<<<< HEAD
    pub fn execute(context: &mut Value, _tx: Option<&Transaction>, _blk: Option<&Block>) -> u8 {
        let hash = context["script"]["hash"].as_str().unwrap_or("");
        let input = context["witness"]["input"].as_str().unwrap_or("");

        if hash.is_empty() || input.is_empty() {
            return 0;
        }

        let computed = crypto::hash(input.as_bytes()).to_hex();
        if computed == hash {
            1
        } else {
            0
        }
=======
    /// Construct default opcode metadata.
    pub fn new() -> Self {
        todo!()
    }

    /// Run CHECKHASH at `node` within `script`.
    pub fn execute(script: &Script, node: &Value) -> u8 {
        todo!()
>>>>>>> d0f828fa (fix: rustscript rust implementation)
    }
}
